src/probnum/diffeq/odesolver.py

```diff
@@ -3,18 +3,103 @@
 """
 
 from abc import ABC, abstractmethod
+import warnings
+
+from probnum.diffeq.odesolution import ODESolution
 
 
 class ODESolver(ABC):
     """
     Interface for ODESolver.
     """
 
-    @abstractmethod
-    def solve(self, ivp, minstep, maxstep, **kwargs):
+    def __init__(self, ivp):
+        self.ivp = ivp
+        self.num_steps = 0
+
+    def solve(self, firststep, steprule, **kwargs):
         """
-        Every ODE solver has a solve() method.
-        Optional: callback function. Allows  e.g. printing variables
-        at runtime.
+        Solve an IVP.
+
+        Parameters
+        ----------
+        firststep : float
+            First step for adaptive step size rule.
+        steprule : StepRule
+            Step-size selection rule, e.g. constant steps or adaptive steps.
+        """
+        t, current_rv = self.initialise()
+        times, rvs = [t], [current_rv]
+        stepsize = firststep
+
+        while t < self.ivp.tmax:
+
+            t_new = t + stepsize
+            proposed_rv, errorest = self.step(t, t_new, current_rv, **kwargs)
+
+            if steprule.is_accepted(stepsize, errorest):
+                self.num_steps += 1
+                self.pre_accepted_callback(
+                    time=t_new, current_guess=proposed_rv, current_error=errorest
+                )
+                t = t_new
+                current_rv = proposed_rv
+                times.append(t)
+                rvs.append(current_rv)
+
+            suggested_stepsize = self._suggest_step(stepsize, errorest, steprule)
+            stepsize = min(suggested_stepsize, self.ivp.tmax - t)
+
+        odesol = self.postprocess(times=times, rvs=rvs)
+        return odesol
+
+    def _suggest_step(self, step, errorest, steprule):
+        """
+        Suggests step according to steprule and warns if
```

Comment:
What a shame that black doesn't do this autoamtically

Reply:
Since I checked Black's changelog earlier: They seem to be working on it and actually have a first implementation for string re-wrapping (accessible through `--experimental-string-processing`), but as the name suggests they're not quite there yet unfortunately^^